discovery-provider/src/tasks/user_replica_set.py

```diff
@@ -0,0 +1,269 @@
+                # Process L2 Content Node operations
+                elif event_type == user_replica_set_manager_event_types_lookup['add_or_update_content_node']:
+                    cnode_record = parse_usrm_cnode_record(
+                        self,
+                        update_task,
+                        session,
+                        entry,
+                        cnode_events_lookup[cnode_sp_id]["content_node"]
+                    )
+                    if cnode_record is not None:
+                        cnode_events_lookup[cnode_sp_id]["content_node"] = cnode_record
+                        cnode_events_lookup[cnode_sp_id]["events"].append(event_type)
+            num_user_replica_set_changes += len(user_events_tx)
+
+    # for each record in user_replica_set_events_lookup, invalidate the old record and add the new record
+    # we do this after all processing has completed so the user record is atomic by block, not tx
+    for user_id, value_obj in user_replica_set_events_lookup.items():
+        logger.info(f"user_replica_set.py | Replica Set Processing Adding {value_obj['user']}")
+        invalidate_old_user(session, user_id)
+        session.add(value_obj["user"])
+
+    for content_node_id, value_obj in cnode_events_lookup.items():
+        logger.info(f"user_replica_set.py | Content Node Processing Adding {value_obj['content_node']}")
+        invalidate_old_cnode_record(session, content_node_id)
+        session.add(value_obj["content_node"])
+
+    return num_user_replica_set_changes, user_ids
+
+# Reconstruct endpoint string from primary and secondary IDs
+# Attempt to retrieve from cached values populated in index_network_peers.py
+# If unavailable, then a fallback to ethereum mainnet contracts will occur
+# Note that in the case of an invalid spID - one that is not yet registered on
+# the ethereum mainnet contracts, there will be an empty value in the returned
+# creator_node_endpoint
+# If this discrepancy occurs, a client replica set health check sweep will
+# result in a client-initiated failover operation to a valid set of replicas
+def get_endpoint_string_from_sp_ids(
+        self,
+        update_task,
+        primary,
+        secondaries,
+        redis
+):
+    sp_factory_inst = None
+    endpoint_string = None
+    primary_endpoint = None
+    try:
+        sp_factory_inst, primary_endpoint = get_endpoint_from_id(
+            self,
+            update_task,
+            sp_factory_inst,
+            primary
+        )
+        endpoint_string = "{}".format(primary_endpoint)
+        for secondary_id in secondaries:
+            secondary_endpoint = None
+            sp_factory_inst, secondary_endpoint = get_endpoint_from_id(
+                self,
+                update_task,
+                sp_factory_inst,
+                secondary_id
+            )
+            if secondary_endpoint:
+                endpoint_string = "{},{}".format(endpoint_string, secondary_endpoint)
```

Comment:
the python3 way to format strings is to use f strings which basically does the same thing as format
```python3
endpoint_string = f"{endpoint_string},{secondary_endpoint}"
```

Reply:
sure